### PR TITLE
[jog_arm] Remove scale/joint parameter

### PR DESCRIPTION
**Pull request by AndyZe**
Previously, the `scale/joint parameter` was also being used to check velocity limits. This was confusing (even to me!). It's also redundant with the SRDF velocity limits.

So that velocity check is removed.

This PR also adds:

-  a unit test for direct joint jogging commands.

- a check that publish_period is >0

### CODE REVIEW

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml [hunk 1]
```diff
@@ -7,9 +7,12 @@ use_gazebo: true # Whether the robot is started in a Gazebo simulation environme
...
 scale:
-  linear:  0.6  # Max linear velocity. Meters per publish_period. Units is [m/s]. Only used if command_in_type=="unitless"
-  rotational:  0.3 # Max angular velocity. Rads per publish_period. Units is [rad/s]. Only used if command_in_type=="unitless"
-  joint: 0.3  # Max joint angular/linear velocity. Rads or Meters per publish period. Units is [rad/s] or [m/s].
+  # Scale parameters are only used if command_in_type=="unitless"
+  linear:  0.6  # Max linear velocity. Meters per publish_period. Unit is [m/s]. Only used for Cartesian commands.
+  rotational:  0.3 # Max angular velocity. Rads per publish_period. Unit is [rad/s]. Only used for Cartesian commands.
+  # Max joint angular/linear velocity. Rads or Meters per publish period. Only used for joint commands on joint_command_in_topic.
+  joint: 10
+max_joint_velocity: 1  # Unit is rad/s. Applies to Cartesian commands.
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: >  Applies to Cartesian commands.

I don't understand this comment - does it not apply to other types of commands? Does this max get applied at the *input* or *output* of JogArm? How is there a joint limit for a Cartesian command?

File: moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_arm_data.h [hunk 2]
```diff
@@ -89,9 +89,9 @@ struct JogArmParameters
 {
   std::string move_group_name, joint_topic, cartesian_command_in_topic, robot_link_command_frame, command_out_topic,
       planning_frame, warning_topic, joint_command_in_topic, command_in_type, command_out_type;
-  double linear_scale, rotational_scale, joint_scale, lower_singularity_threshold, hard_stop_singularity_threshold,
-      collision_proximity_threshold, low_pass_filter_coeff, publish_period, incoming_command_timeout,
-      joint_limit_margin, collision_check_rate;
+  double linear_scale, rotational_scale, joint_scale, max_joint_velocity, lower_singularity_threshold,
+      hard_stop_singularity_threshold, collision_proximity_threshold, low_pass_filter_coeff, publish_period,
+      incoming_command_timeout, joint_limit_margin, collision_check_rate;
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_arm_data.h]: I would prefer these vars one per line, as Stack Overflow says:

> If you declare one variable/object per line not only does it solve this problem, but it makes the code clearer and prevents silly mistakes when declaring pointers.

https://stackoverflow.com/questions/6838408/how-can-i-declare-and-define-multiple-variables-in-one-line-using-c/6838480

File: moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_calcs.h [hunk 3]
```diff
@@ -127,7 +127,8 @@ class JogCalcs
   // For jacobian calculations
   Eigen::MatrixXd jacobian_, pseudo_inverse_, matrix_s_;
   Eigen::JacobiSVD<Eigen::MatrixXd> svd_;
-  Eigen::VectorXd delta_theta_;
+  // Use ArrayXd type to enable more coefficient-wise operations
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_calcs.h]: :heart: comment

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 4]
```diff
@@ -541,15 +541,16 @@ double JogCalcs::decelerateForSingularity(const Eigen::VectorXd& commanded_veloc
...
+  if ((maximum_joint_change / parameters_.publish_period) > parameters_.max_joint_velocity)
   {
     // Scale the entire joint velocity vector so that each joint velocity is below min, and the output movement is
-    // scaled uniformly to match expected motion
-    calculated_joint_change =
-        calculated_joint_change * parameters_.joint_scale * parameters_.publish_period / maximum_joint_change;
+    // scaled uniformly to match direction of expected motion.
+    // The formula is delta_j(i) = delta_j(i) / (|max_velocity| / velocity_limit)
+    calculated_joint_change = calculated_joint_change /
+                              ((maximum_joint_change / parameters_.publish_period) / parameters_.max_joint_velocity);
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Are there checks that all these variables are non-zero? Asserts at top of function would be nice

File: moveit_experimental/moveit_jog_arm/src/jog_interface_base.cpp [hunk 5]
```diff
@@ -101,6 +102,12 @@ bool JogInterfaceBase::readParameters(ros::NodeHandle& n)
   rosparam_shortcuts::shutdownIfError(parameter_ns, error);
 
   // Input checking
+  if (ros_parameters_.max_joint_velocity < 0.)
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/src/jog_interface_base.cpp]:
```suggestion
  if (ros_parameters_.max_joint_velocity <= 0.)
```

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 6]
```diff
@@ -541,15 +541,16 @@ double JogCalcs::decelerateForSingularity(const Eigen::VectorXd& commanded_veloc
   return velocity_scale;
 }
 
-void JogCalcs::enforceJointVelocityLimits(Eigen::VectorXd& calculated_joint_change)
+void JogCalcs::enforceJointVelocityLimits(Eigen::ArrayXd& calculated_joint_change)
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: There should be an associated unit test to verify this math makes sense, please add

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml (hunk 1 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: OK, added more/better comments

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 4 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Added a check for publish_period > 0 in jog_interface_base.cpp.

max_joint_velocity already got checked.

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml (hunk 1 above)
Comment (davetcoleman) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: I'm still confused about how you apply a joint limit to a Cartesian commands though...

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 6 above)
Comment (davetcoleman) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: This wasn't resolved...

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml (hunk 1 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: delta_theta = J^t * delta_x

So it's after you calculate a delta_theta

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml [hunk 7]
```diff
@@ -7,9 +7,15 @@ use_gazebo: true # Whether the robot is started in a Gazebo simulation environme
...
-  joint: 0.3  # Max joint angular/linear velocity. Rads or Meters per publish period. Units is [rad/s] or [m/s].
+  # Scale parameters are only used if command_in_type=="unitless"
+  linear:  0.6  # Max linear velocity. Meters per publish_period. Unit is [m/s]. Only used for Cartesian commands.
+  rotational:  0.3 # Max angular velocity. Rads per publish_period. Unit is [rad/s]. Only used for Cartesian commands.
+  # Max joint angular/linear velocity. Rads or Meters per publish period. Only used for joint commands on joint_command_in_topic.
+  joint: 10
+# Joint velocity limit on outgoing command to robot joints.
+# Unit is rad/s. This velocity limit is checked before the command is published from jog_arm.
+# Note: joints likely have additional velocity limits in the SRDF that are also checked.
+max_joint_velocity: 1
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: So this is different than ``max_velocity_scaling_factor`` that we use elsewhere in MoveIt?

And different than the ``joint_limits.yaml`` file we use elsewhere in MoveIt (via SRDF, as you mentioned)?

Does this mean that no joint can exceed this one value? What if some joints are revolute and others prismatic? This single variable seems short-sighted.

File: moveit_experimental/moveit_jog_arm/src/jog_interface_base.cpp [hunk 8]
```diff
@@ -101,6 +102,18 @@ bool JogInterfaceBase::readParameters(ros::NodeHandle& n)
   rosparam_shortcuts::shutdownIfError(parameter_ns, error);
 
   // Input checking
+  if (ros_parameters_.publish_period <= 0.)
+  {
+    ROS_WARN_NAMED(LOGNAME, "Parameter 'max_joint_velocity' should be "
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/src/jog_interface_base.cpp]:
```suggestion
    ROS_WARN_NAMED(LOGNAME, "Parameter 'publish_period' should be "
```

File: moveit_experimental/moveit_jog_arm/test/integration/test_jog_arm_integration.py [hunk 9]
```diff
@@ -29,11 +29,11 @@ class JointJogCmd(object):
     def __init__(self):
         self._pub = rospy.Publisher(JOINT_JOG_COMMAND_TOPIC, JointJog, queue_size=1)
 
-    def send_cmd(self, joint_pos):
+    def send_joint_velocity_cmd(self, joint_pos):
```
Comment (davetcoleman) [inline moveit_experimental/moveit_jog_arm/test/integration/test_jog_arm_integration.py]: nice cleanup!

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml (hunk 7 above)
Comment (davetcoleman) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: @henningkayser thoughts?

Comment (davetcoleman) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: @henningkayser thoughts?

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: I would actually be happy to drop this parameter and just check the SRDF joint velocity limits.

I want to keep some of the other functionality from this PR, though, like the updated integration test.

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: I guess the only argument for keeping that parameter in this PR is if the user wanted to launch jogging for a very slow, delicate motion (or something like that) without modifying the robot SRDF.

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml [hunk 10]
```diff
@@ -7,9 +7,11 @@ use_gazebo: true # Whether the robot is started in a Gazebo simulation environme
 robot_link_command_frame:  world  # commands must be given in the frame of a robot link. Usually either the base or end effector
 command_in_type: "speed_units" # "unitless"> in the range [-1:1], as if from joystick. "speed_units"> cmds are in m/s and rad/s
 scale:
-  linear:  0.6  # Max linear velocity. Meters per publish_period. Units is [m/s]. Only used if command_in_type=="unitless"
-  rotational:  0.3 # Max angular velocity. Rads per publish_period. Units is [rad/s]. Only used if command_in_type=="unitless"
-  joint: 0.3  # Max joint angular/linear velocity. Rads or Meters per publish period. Units is [rad/s] or [m/s].
+  # Scale parameters are only used if command_in_type=="unitless"
+  linear:  0.6  # Max linear velocity. Meters per publish_period. Unit is [m/s]. Only used for Cartesian commands.
+  rotational:  0.3 # Max angular velocity. Rads per publish_period. Unit is [rad/s]. Only used for Cartesian commands.
+  # Max joint angular/linear velocity. Rads or Meters per publish period. Only used for joint commands on joint_command_in_topic.
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: I'm now confused by this. I would expect this to be a velocity limit in rads or meters per second and not per publish period. Also, 10 is a really high example value here, no? Or does it count for all joints together...

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 11]
```diff
@@ -541,20 +533,22 @@ double JogCalcs::decelerateForSingularity(const Eigen::VectorXd& commanded_veloc
...
+
+      // Apply velocity bounds
+    }
   }
-}
+  ROS_WARN_STREAM("===");
 
-bool JogCalcs::checkIfJointsWithinURDFBounds(trajectory_msgs::JointTrajectory& new_joint_traj)
-{
+/*
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Please add a NOTE/TODO why this is commented

File: moveit_experimental/moveit_jog_arm/test/integration/test_jog_arm_integration.py [hunk 12]
```diff
@@ -72,17 +71,69 @@ def test_jog_arm_generates_joint_trajectory_when_joint_jog_command_is_received(n
...
-    rospy.sleep(test_duration)
-    # test_duration/publish_period is the expected number of messages in this duration.
+    rospy.sleep(TEST_DURATION)
+    # TEST_DURATION/PUBLISH_PERIOD is the expected number of messages in this duration.
+    # Allow a small +/- window due to rounding/timing errors
+    assert len(received) >= TEST_DURATION/PUBLISH_PERIOD - 5
+    assert len(received) <= TEST_DURATION/PUBLISH_PERIOD + 5
+
+
+def test_jog_arm_joint_command(node):
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/test/integration/test_jog_arm_integration.py]: Nice test

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml (hunk 7 above)
Comment (henningkayser) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: @davetcoleman I totally relate to your confusion (see comment further below) ;)

> I guess the only argument for keeping that parameter in this PR is if the user wanted to launch jogging for a very slow, delicate motion (or something like that) without modifying the robot SRDF.

I see.. how about JogArm just looks for max_acceleration_scaling_factor and max_velocity_scaling_factor parameters?

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 11 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Yeah, this is Work In Progress. Which I think we discussed reverting, submitting the basics of this PR, then adding a followup PR. Sorry for not marking WIP

File: moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml (hunk 10 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/config/ur_simulated_config.yaml]: I wouldn't worry about it now, it's only used for "unitless" commands like from a joystick. They have to be scaled somehow.

Yeah, 10 is high. I'll change it to 1.

**Comment by AndyZe** [comment]
OK, I squashed the commits and simplified the scope of this PR to removing the redundant joint velocity check.

**Comment by davetcoleman** [comment]
Please remove WIP

**Comment by davetcoleman** [comment]
I don't know why codecov failed, but Andy wrote more unit tests so this is a good PR regarding testing. Merging.

**Pull request merged**